Alex_RTOS_Sailboat/RTOS_framework_v1/src/sail_gps.c

```diff
@@ -0,0 +1,344 @@
+ * Created: 2022-06-02 8:04 PM
+ *  Author: secul
+ */
+
+#include "sail_gps.h"
+#include <stdio.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include "FreeRTOS.h"
+#include "FreeRTOSConfig.h"
+#include "task.h"
+#include "sail_ctrl.h"
+#include "sail_tasksinit.h"
+#include "usart_interrupt.h"
+
+#include "sail_debug.h"
+
+#define HEADER_FMT				"DALSAIL,%03"PRIu16
+#define FIELD_FMT				",%"PRIi32
+
+#define RADIO_BUFFER_LENGTH		NMEA_BUFFER_LENGTH
+
+static bool init_flag = false;
+static char msg_buffer[RADIO_BUFFER_LENGTH];
+
+//stores all the msg types of the gps sensor
+volatile GPS_AllMsgs GPS_data;
+
+void ReadGPS(void) {
+	DEBUG_Write("Reading GPS...\r\n");
+	uint16_t loop_cnt = 0;
+	//set msg type sum to 0 since no messages processed yet
+	GPS_data.msg_type_sum = 0;
+
+	// Event bits for holding the state of the event group
+	EventBits_t event_bits;
+
+	TickType_t read_gps_delay = pdMS_TO_TICKS(GPS_SLEEP_PERIOD_MS);
+
+	while (1) {
+
+		event_bits = xEventGroupWaitBits(mode_event_group,                        /* Test the mode event group */
+			CTRL_MODE_AUTO_BIT | CTRL_MODE_REMOTE_BIT, /* Wait until the sailboat is in AUTO or REMOTE mode */
+			pdFALSE,                                 /* Bits should not be cleared before returning. */
+			pdFALSE,                                 /* Don't wait for both bits, either bit will do. */
+			portMAX_DELAY);                          /* Wait time does not expire */
+			
+		taskENTER_CRITICAL();
+		watchdog_counter |= 0x01;
+		taskEXIT_CRITICAL();
+
+		GPS_On();
+
+		running_task = eReadWeatherSensor; //Replace this with gps, see tasksinit.c/.h
+
+		NMEA_GenericMsg msg;
+
+		if (GPS_RxMsg(&msg) == STATUS_OK) {
+			loop_cnt++;
+			//DEBUG_Write("\nStatus OK\n");
+			//DEBUG_Write("loop count %d\r\n", loop_cnt);
+			//store msg into msg array at index corresponding to msg type
+			GPS_data.msg_array[msg.type] = msg;
+			//add type to msg type sum to keep track of the saved messages
+			GPS_data.msg_type_sum += msg.type;
+
+			DEBUG_Write("\nAll fields obtained... Going to sleep...\r\n");
+			GPS_data.msg_type_sum = 0;
+
+			//store weather station data into appropriate structs
+			assign_gps_readings();
+			//check if waypoint was reached and affect as necessary
+			DEBUG_Write("checking waypoint...\r\n");
+			check_waypoint_state();
+
+			//Necessary??????????????
+			DEBUG_Write("processing wind...\r\n");
+			//calculate wind parameters
+			process_wind_readings();
+			DEBUG_Write("processing heading...\r\n");
+			//calculate heading parameters
+			process_heading_readings();
+
+			//put thread to sleep until a specific tick count is reached
+			vTaskDelay(read_gps_delay);
+		}
+	}
+}
+
+void GPS_Sleep_Sec(unsigned time_sec) {
+	GPS_Disable();
+	//put thread to sleep for number of ticks specified
+	vTaskDelay(time_sec * configTICK_RATE_HZ);
+}
+
+void GPS_On(void) {
+	GPS_Init();
+	GPS_Enable();
+}
+
+enum status_code GPS_Init(void)
```

Comment:
Great Error Handling!

Reply:
Hopefully it actually works. We should probably try and add the same format of error handling where we can. Will probably be helping when doing testing later on